ginga/misc/plugins/SaveImage.py

```diff
@@ -0,0 +1,477 @@
+            yield key
+
+    def _format_extname(self, ext):
+        """Pretty print given extension name and number tuple."""
+        if ext is None:
+            outs = ext
+        else:
+            outs = '{0},{1}'.format(ext[0], ext[1])
+        return outs
+
+    def browse_outdir(self):
+        """Browse for output directory."""
+        self.dirsel.popup(
+            'Select directory', self.w.outdir.set_text, initialdir=self.outdir)
+        self.set_outdir()
+
+    def set_outdir(self):
+        """Set output directory."""
+        dirname = self.w.outdir.get_text()
+        if os.path.isdir(dirname):
+            self.outdir = dirname
+            self.logger.debug('Output directory set to {0}'.format(self.outdir))
+        else:
+            self.w.outdir.set_text(self.outdir)
+            self.logger.error('{0} is not a directory'.format(dirname))
+
+    def set_suffix(self):
+        """Set output suffix."""
+        self.suffix = self.w.suffix.get_text()
+        self.logger.debug('Output suffix set to {0}'.format(self.suffix))
+
+    def _write_history(self, pfx, hdu, linechar=60, indentchar=2):
+        """Write change history to given HDU header.
+        Limit each HISTORY line to given number of characters.
+        Subsequent lines of the same history will be indented.
+        """
+        if self.history_obj is None:
+            return
+
+        file_dict = self.history_obj.name_dict[self.chname]
+        chistory = []
+        ind = ' ' * indentchar
+
+        # NOTE: List comprehension too slow!
+        for key in file_dict:
+            if not key.startswith(pfx):
+                continue
+            for bnch in itervalues(file_dict[key]):
+                chistory.append('{0} {1}'.format(bnch.MODIFIED, bnch.DESCRIP))
+
+        # Add each HISTORY prettily into header, sorted by timestamp
+        for s in sorted(chistory):
+            for i in range(0, len(s), linechar):
+                subs = s[i:i+linechar]
+                if i > 0:
+                    subs = ind + subs.lstrip()
+                hdu.header.add_history(subs)
+
+    def _write_header(self, image, hdu):
+        """Write header from image object to given HDU."""
+        hduhdr = hdu.header
+
+        # Ginga image header object for the given extension only.
+        # Cannot use get_header() because that might also return PRI hdr.
+        ghdr = image.metadata['header']
+
+        for key in ghdr:
+            # Need this to avoid duplication because COMMENT is a weird field
+            if key.upper() == 'COMMENT':
+                continue
+
+            bnch = ghdr.get_card(key)
+
+            # Insert new keyword
+            if key not in hduhdr:
+                hduhdr[key] = (bnch.value, bnch.comment)
+
+            # Update existing keyword
+            elif hduhdr[key] != bnch.value:
+                hduhdr[key] = bnch.value
+
+    def _write_mosaic(self, key, outfile):
```

Comment:
I thought about moving I/O related methods to `ginga.util.io_fits` module but decided against it because the functionality here also depends on other plugins (e.g., inserting history from `ChangeHistory` plugin), so things get a bit tricky to generalize.